### PR TITLE
F# Frameworks

### LINKED ISSUE #2869: [F#] Add frank
Hi @panesofglass,

Would you mind to add (or consent that we add) [frank](https://github.com/frank-fs/frank) in `here` ?

This project compare performance based on 3 routes :

+ `GET` `/`, **SHOULD** return a successful status code and an empty string
+ `GET` `/user/:id`, **SHOULD** return a successful status code and the id
+ `POST` `/user`, **SHOULD** return a successful status code and an empty string

Regards,

### PR DESCRIPTION
**Pull request by panesofglass**
- [x] Add [Frank](https://github.com/frank-fs/frank) #2869 
- [x] Add [Falco](https://github.com/pimbrouwers/Falco)
- [x] Add [Giraffe](https://github.com/giraffe-fsharp/Giraffe)

------------------
Closes #2869

### CODE REVIEW

**Comment by pimbrouwers** [comment]
The Falco code looks perfect. Thank you for doing this Ryan!

**Comment by panesofglass** [comment]
@waghanza, looks like it doesn’t like the sorbet publish command. I’ll look into it.

**Comment by panesofglass** [comment]
It seems the WebSharper build only works out of the box on netcoreapp2.1. I'll comment this one out for now and let @granicz or @Tarmil submit a fix.

File: fsharp/falco/config.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+framework:
+  website: github.com/pimbrouwers/Falco
```
Comment (waghanza) [inline fsharp/falco/config.yaml:2]: rather use `github: pimbrouwers/Falco` instead

File: fsharp/falco/config.yaml [hunk 2]
```diff
@@ -0,0 +1,3 @@
...
+  version: 1.0.10-alpha
```
Comment (waghanza) [inline fsharp/falco/config.yaml:3]: I prefer to use **stable** release, no `alpha`

@pimbrouwers I found only `alpha` on https://www.nuget.org/packages/Falco

**Comment by pimbrouwers** [comment]
Great work Ryan! Looking forward to seeing the results!

**Comment by panesofglass** [comment]
@waghanza any idea why the git pull failed on the Travis builds? Is this due to some sort of request throttling? Any chance you can re-run the build?

**Comment by waghanza** [comment]
@panesofglass Sometimes, it happens ... seems that a network issue or git was ammended, btw, I've restarted (PS : In the next version we will use https://the-benchmarker.semaphoreci.com/projects/web-frameworks)

@pimbrouwers Is there any **stable** release of `Falco` ?

**Comment by waghanza** [comment]
failed at https://travis-ci.com/github/the-benchmarker/web-frameworks/jobs/344675833#L399  @panesofglass  @Tarmil any idea ?

**Comment by pimbrouwers** [comment]
@waghanza The latest alpha build is stable and should be officially released shortly. There won't be any core code changes during the release. Only improved testing. The current version is ready for production.

**Comment by waghanza** [comment]
Ok, then I'll merge this time, when someone with `F#` skill could fix `CI`

However, I prefer **NOT** to include in this list any version not labelled as stable :stuck_out_tongue:

**Comment by pimbrouwers** [comment]
@waghanza can you wait an hour or so? My plan was to release this weekend. So I can just expedite?

**Comment by waghanza** [comment]
sure ... In fact, I think, it will take time to fix `CI` :stuck_out_tongue: 

but, I'll wait until you release, at least this project is also to promote OSS :heart: and I prefer to promote anything stable :+1:

**Comment by pimbrouwers** [comment]
No problem. Makes sense. I'll do my best to get it done today.

**Comment by waghanza** [comment]
No hurry @pimbrouwers, I'll promote mostly after https://github.com/the-benchmarker/web-frameworks/pull/2593#partial-pull-merging is finished, cause this will run on `cloud` infrastructure => https://digitalocean.com

**Comment by pimbrouwers** [comment]
Ok! v1.1.0 has been deployed to nuget. @panesofglass do you mind changing the package version to 1.1.0 in the proj file? Or would you rather I submit a PR on your PR?

**Comment by panesofglass** [comment]
@waghanza WebSharper has a known issue running when targeting netcoreapp3.1 on docker. Should I pull that to another branch to track separately?

**Comment by panesofglass** [comment]
@pimbrouwers Done!

@waghanza I took websharper out of the build list. That should get this to pass. We can add it again when they have it fixed. See https://github.com/dotnet-websharper/core/issues/1044#issuecomment-639575833

**Comment by pimbrouwers** [comment]
Thanks Ryan! I submitted a PR as well. But shouldn't be a game changer, so feel free to ignore. If we can roll it in, that would be awesome.

File: fsharp/falco/config.yaml (hunk 2 above)
Comment (waghanza) [reply fsharp/falco/config.yaml:3]: You forget to update this. Moreover, I prefer having `X.Y` format -> **1.1**

File: fsharp/frank/web.fsproj [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="Program.fs" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="Frank" Version="6.0.0" />
```
Comment (waghanza) [inline fsharp/frank/web.fsproj:9]: It's better to use `6.0.*` to have **6.0.x**

File: fsharp/falco/web.fsproj [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="Program.fs" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="Falco" Version="1.1.0" />
```
Comment (waghanza) [inline fsharp/falco/web.fsproj:9]: It's better to use `1.1.*` to have **1.1.x**

File: fsharp/websharper/web.fsproj [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    <WebProjectOutputDir>$(MSBuildThisFileDirectory)/wwwroot</WebProjectOutputDir>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Compile Include="Main.fs" />
+    <None Include="wsconfig.json" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="WebSharper" Version="4.6.3.388" />
```
Comment (waghanza) [inline fsharp/websharper/web.fsproj:15]: It's better to use `4.6.*` to have **4.6.x**

File: fsharp/websharper/web.fsproj [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    <PackageReference Include="WebSharper.AspNetCore" Version="4.6.1.129" />
```
Comment (waghanza) [inline fsharp/websharper/web.fsproj:16]: It's better to use `4.6.*` to have **4.6.x**

File: fsharp/websharper/web.fsproj [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+  <ItemGroup>
+    <Compile Include="Main.fs" />
+    <None Include="wsconfig.json" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="WebSharper" Version="4.6.3.388" />
+    <PackageReference Include="WebSharper.AspNetCore" Version="4.6.1.129" />    
+    <PackageReference Include="WebSharper.FSharp" Version="4.6.3.388" />
```
Comment (waghanza) [inline fsharp/websharper/web.fsproj:17]: It's better to use `4.6.*` to have **4.6.x**

File: fsharp/websharper/web.fsproj [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    <PackageReference Include="WebSharper.UI" Version="4.6.3.210" />
```
Comment (waghanza) [inline fsharp/websharper/web.fsproj:18]: It's better to use `4.6.*` to have **4.6.x**

**Comment by pimbrouwers** [comment]
I submitted a PR to Ryan's repo that will take care of the yaml file. I'm
not a fan of floating versions. But if you'd rather have the wildcard in
there, that's fine too.

File: fsharp/websharper/Main.fs [hunk 9]
```diff
@@ -0,0 +1,48 @@
+namespace web
```
[the remark this answers is not in the document]
Comment (panesofglass) [reply fsharp/websharper/Main.fs:1]: Will do

**Comment by waghanza** [comment]
@pimbrouwers I do not know if there is an other way in `F#` to define a version range

**Comment by panesofglass** [comment]
I don't suppose I need to worry about the AnyLint failure? I've never even heard of that plugin before.

**Comment by waghanza** [comment]
Nop, it's good. `AnyLint` complains about **markdown**, I'll disable it (since it will `README.md` will be completely rewritten)

**Comment by panesofglass** [comment]
@waghanza looks like the last build had the same issue with pulling and merging the branch. :( Any chance you could kick off another?

**Comment by waghanza** [comment]
sure. I'll test locally, generate some result, and push again, this will trigger `CI` again. I'm having more and more build problems with actual `CI`

**Comment by waghanza** [comment]
I suppose the `CI` fails because of some merge strategy (we are using **squash**) different in your repo @panesofglass (github's default is to create a merge commit)

**Comment by pimbrouwers** [comment]
Shouldn't these be appearing in the results? It says they were update 2020/06/06.

**Comment by waghanza** [comment]
The results are updated in https://github.com/panesofglass/web-frameworks/tree/fsharp/frank#results, the process is still manual, and takes a long long time

**Comment by waghanza** [comment]
@panesofglass @pimbrouwers @dustinmoris As to have a professional benchmark, I would like to add you as `codeowners` (of the framework you make / contribute to). Do you consent on such a thing ?

**Comment by dustinmoris** [comment]
Fine with me! Thanks

**Pull request merged**